properties/test_pandas_roundtrip.py

```diff
@@ -0,0 +1,91 @@
+"""
+Property-based tests for roundtripping between xarray and pandas objects.
+"""
+import hypothesis.extra.numpy as npst
+import hypothesis.extra.pandas as pdst
+import hypothesis.strategies as st
+from hypothesis import given
+
+import numpy as np
+import pandas as pd
+import xarray as xr
+
+numeric_dtypes = st.one_of(
+    npst.unsigned_integer_dtypes(), npst.integer_dtypes(), npst.floating_dtypes()
+)
+
+numeric_series = numeric_dtypes.flatmap(lambda dt: pdst.series(dtype=dt))
+
+an_array = npst.arrays(
+    dtype=numeric_dtypes,
+    shape=npst.array_shapes(max_dims=2),  # can only convert 1D/2D to pandas
+)
+
+
+@st.composite
+def datasets_1d_vars(draw):
+    """Generate datasets with only 1D variables
+
+    Suitable for converting to pandas dataframes.
+    """
+    n_vars = draw(st.integers(min_value=1, max_value=3))
+    n_entries = draw(st.integers(min_value=0, max_value=100))
+    dims = ("rows",)
+    vars = {}
+    for _ in range(n_vars):
```

Comment:
This pattern - draw a number, then draw that many elements - is tempting but [tends to be inefficient](https://github.com/HypothesisWorks/hypothesis/blob/master/guides/strategies-that-shrink.rst) when Hypothesis tries to minimse any failures.

The alternative, which we recommend, is to generate collections using the `st.lists()` strategy - that way Hypothesis will be able to operate in terms of elements of the list.

In this case it's probably only worth doing so for either the vars or entries dimension, and keep the other as-is.  If you're keen to do both, it's complicated enough that I'd just fall back on the Hypothesis pandas extension and `.map(pd.DataFrame.to_xarray)` over the result :sweat_smile:

Reply:
You could draw `indices` before the loop, then draw a list of (name, array) tuples.  Or in this case you could use `st.dictionaries()` to, well, generate a list of key-value tuples internally.

The other nice trick would be to draw your index first, and use it's length - deleting elements from that will be *slightly* more efficient than shrinking the `n_elements` parameter.

Putting it all togther, I'd write 

```python
idx = draw(pdst.indexes(dtype="u8", min_size=0, max_size=100))
vars_strat = st.dictionaries(
    keys=st.text(),
    values=npst.arrays(dtype=numeric_dtypes, shape=len(idx)).map(partial(xr.Variable, ("rows",))),
    min_size=1,
    max_size=3,
)
return xr.Dataset(draw(vars_strat), coords={"rows": idx})
```

Reply:
Thanks, that does look neater!